### PR TITLE
Enable roberta embedding

### PR DESCRIPTION
**Pull request by yeonsily**
We set position_ids and input_ids as [batch_size, bucket_size] on hpu so need to modify the current roberta embedding forward function.

e.g. position_id on gpu
[0,1,2,3,4,5,6,7]
but position_id on hpu
[0,1,2,3,4,5,6,7,0,0,0,.....0,0] which size is 128 with padding in this case.

One thing I noticed is torch.equal() on hpu is not working properly and I have to run it on cpu.
That code is nothing but checking pre-condition but we will need to investigate it.

This PR has a dependency on https://github.com/HabanaAI/vllm-fork/pull/758

Update-
We found that the issue is from dry_run=True in hpu_graph. I think we should disable it.

From pt-integration team,

```
`disable_tensor_cache=True` is an experimental feature and should be used with caution. While some models can work with this flag directly, others need to pass the cache_tensors_list alongside disable_tensor_cache=True to ensure that certain tensors are not freed.
When `disable_tensor_cache=True` is enabled, it will free all tensors except user outputs and view tensors. However, certain tensors, such as in-place tensors, cannot be identified by hpugraph and will also be freed. To prevent this, these tensors must be specified in the cache_tensors_list.
Also, when `disable_tensor_cache=True` is enabled, dry_run is enabled by default.
when dry run is enabled, the intermediate values/SingleHpugraphs are not evaluated until the full graph is captured, so the necessary marksteps/evaluation happening internally will not update the values. eg: local_scalar_dense
So this can lead to accuracy issues.
```

Update2-
'disable_tensor_cache' is configurable by `PT_HPUGRAPH_DISABLE_TENSOR_CACHE`. I reverted my change and commented it to README.

### CODE REVIEW

**Comment by afierka-intel** [comment]
@yeonsily thank you for the PR.

Rebase on latest habana_main branch to fix known issues in CI. Then ask for review again.

Thank you in advance!

**Comment by yeonsily** [comment]
> @yeonsily thank you for the PR.
> 
> Rebase on latest habana_main branch to fix known issues in CI. Then ask for review again.
> 
> Thank you in advance!

Done. Thank you!

**Comment by yeonsily** [comment]
I see that it failed to run tests/lora/test_llama_hpu.py::test_llama_lora_1x with graph compile error.
But I tried the same test on 1.19 docker on local machine and it passed.
I'm not sure if that's some setup specific issue.

Other failed test are because of hccl or device acq error.

File: vllm/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -857,9 +857,10 @@ def _add_dummy_seq(self, seq_group_metadata_list, is_prompt):
 
     def _maybe_wrap_in_hpu_graph(self, *args, **kwargs):
         return htorch.hpu.wrap_in_hpu_graph(
-            HpuModelAdapter(*args, **kwargs), disable_tensor_cache=True
-        ) if htorch.utils.internal.is_lazy() else HpuModelAdapter(
-            *args, **kwargs)
+            HpuModelAdapter(*args, **kwargs),
+            disable_tensor_cache=True,
+            dry_run=False) if htorch.utils.internal.is_lazy(
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:862]: This is rather a major change if we are setting dry run to False now, without giving any configurability. Let's talk offline on how that affects performance.

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: README_GAUDI.md [hunk 2]
```diff
@@ -366,7 +366,7 @@ Additionally, there are HPU PyTorch Bridge environment variables impacting vLLM
 
 - `PT_HPU_LAZY_MODE`: if `0`, PyTorch Eager backend for Gaudi will be used, if `1` PyTorch Lazy backend for Gaudi will be used. `1` is the default.
 - `PT_HPU_ENABLE_LAZY_COLLECTIVES` must be set to `true` for tensor parallel inference with HPU Graphs.
-- `PT_HPUGRAPH_DISABLE_TENSOR_CACHE` must be set to `false` for llava model.
+- `PT_HPUGRAPH_DISABLE_TENSOR_CACHE` must be set to `false` for llava, roberta models.
```
Comment (michalkuligowski) [inline README_GAUDI.md:369]: "for llava and roberta models"

File: docs/source/getting_started/installation/ai_accelerator/hpu-gaudi.inc.md [hunk 3]
```diff
@@ -363,7 +363,7 @@ Additionally, there are HPU PyTorch Bridge environment variables impacting vLLM
 
 - `PT_HPU_LAZY_MODE`: if `0`, PyTorch Eager backend for Gaudi will be used, if `1` PyTorch Lazy backend for Gaudi will be used. `1` is the default.
 - `PT_HPU_ENABLE_LAZY_COLLECTIVES` must be set to `true` for tensor parallel inference with HPU Graphs.
-- `PT_HPUGRAPH_DISABLE_TENSOR_CACHE` must be set to `false` for llava model.
+- `PT_HPUGRAPH_DISABLE_TENSOR_CACHE` must be set to `false` for llava, roberta models.
```
Comment (michalkuligowski) [inline docs/source/getting_started/installation/ai_accelerator/hpu-gaudi.inc.md:366]: "for llava and roberta models"

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/worker/hpu_model_runner.py (hunk 1 above)
Comment (yeonsily) [reply vllm/worker/hpu_model_runner.py]: done

File: README_GAUDI.md (hunk 2 above)
Comment (yeonsily) [reply README_GAUDI.md]: done

File: README_GAUDI.md [hunk 4]
```diff
@@ -366,7 +366,7 @@ Additionally, there are HPU PyTorch Bridge environment variables impacting vLLM
 
 - `PT_HPU_LAZY_MODE`: if `0`, PyTorch Eager backend for Gaudi will be used, if `1` PyTorch Lazy backend for Gaudi will be used. `1` is the default.
 - `PT_HPU_ENABLE_LAZY_COLLECTIVES` must be set to `true` for tensor parallel inference with HPU Graphs.
-- `PT_HPUGRAPH_DISABLE_TENSOR_CACHE` must be set to `false` for llava model.
+- `PT_HPUGRAPH_DISABLE_TENSOR_CACHE` must be set to `false` for llava and roberta models.
```
Comment (michalkuligowski) [inline README_GAUDI.md:369]: There was still one test suite failing, i rerun it in https://github.com/HabanaAI/vllm-fork/actions/runs/13549640267/job/37870115757?pr=786

File: vllm/model_executor/models/roberta.py [hunk 5]
```diff
@@ -123,6 +122,7 @@ def forward(
 # Adapted from transformers
 def create_position_ids_from_input_ids(input_ids,
                                        padding_idx,
+                                       seq_len,
```
Comment (michalkuligowski) [inline vllm/model_executor/models/roberta.py:125]: Are you planning to upstream it, I am not sure about extending this method signature here cc @kzawora-intel

Comment (yeonsily) [reply vllm/model_executor/models/roberta.py:125]: @michalkuligowski @kzawora-intel Please advise me how you want to handle this.

Comment (michalkuligowski) [reply vllm/model_executor/models/roberta.py:125]: You shouldnt modify model, but rather calling methods, see how llava implements multimodality

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/model_executor/models/roberta.py (hunk 5 above)
Comment (michalkuligowski) [reply vllm/model_executor/models/roberta.py:125]: hi @yeonsily is this PR still relevant or can be closed?

**Comment by libinta** [comment]
@michalkuligowski  this PR is needed as customer requests it also, we will work on it next week.

**Comment by yeonsily** [comment]
@michalkuligowski For "how sequence lenght requirements are implemented for example for alibi or llava", can you please point me out the code location?

File: vllm/model_executor/models/roberta.py [hunk 6]
```diff
@@ -89,23 +89,22 @@ def forward(
...
         token_list = []
-        offset = 0
-        for seq_len in seq_lens:
-            pos_list.append(position_ids[offset:offset + seq_len])
-            token_list.append(input_ids[offset:offset + seq_len])
-            offset += seq_len
-
-        new_pos_list = []
-        for positions, tokens in zip(pos_list, token_list):
+        for offset in range(position_ids.size()[0]):
```
Comment (michalkuligowski) [inline vllm/model_executor/models/roberta.py:92]: @yeonsily  Yes, here is the CustomOp implementation https://github.com/HabanaAI/vllm-fork/blob/222b7f887e9b3d0c828082aac52e1f70cd62d323/vllm/model_executor/custom_op.py#L14
YOu can use it to implement forward_hpu() if the behaviour is different from the native version

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

**Comment by yeonsily** [comment]
Please check https://github.com/HabanaAI/vllm-fork/pull/996 for custom op change.
